runtime/src/mem/jemalloc/mem-jemalloc.c

```diff
@@ -70,46 +72,79 @@ static inline void* alignHelper(void* base_ptr, size_t offset, size_t alignment)
 
 
 // Our chunk replacement hook for allocations (Essentially a replacement for
-// mmap/sbrk.) Grab memory out of the shared heap and give it to jemalloc.
+// mmap/sbrk.) Grab memory out of the fixed shared heap or get an extension
+// chunk, and give it to jemalloc.
 static void* chunk_alloc(void *chunk, size_t size, size_t alignment, bool *zero, bool *commit, unsigned arena_ind) {
 
   void* cur_chunk_base = NULL;
-  size_t cur_heap_size;
 
-  // this function can be called concurrently and it looks like jemalloc
-  // doesn't call it inside a lock, so we need to protect it ourselves
-  pthread_mutex_lock(&heap.alloc_lock);
+  if (heap.isFixed) {
+    //
+    // Get more space out of the fixed heap.
+    //
+    size_t cur_heap_size;
+
+    // this function can be called concurrently and it looks like jemalloc
+    // doesn't call it inside a lock, so we need to protect it ourselves
+    pthread_mutex_lock(&heap.alloc_lock);
+
+    // compute our current aligned pointer into the shared heap
+    //
+    //   jemalloc 4.5.0 man: "The alignment parameter is always a power of two
+    //   at least as large as the chunk size."
+    cur_chunk_base = alignHelper(heap.base, heap.cur_offset, alignment);
+
+    // jemalloc 4.5.0 man: "If chunk is not NULL, the returned pointer must be
+    // chunk on success or NULL on error"
+    if (chunk && chunk != cur_chunk_base) {
+      pthread_mutex_unlock(&heap.alloc_lock);
+      return NULL;
+    }
 
-  // compute our current aligned pointer into the shared heap
-  //
-  //   jemalloc 4.5.0 man: "The alignment parameter is always a power of two at
-  //   least as large as the chunk size."
-  cur_chunk_base = alignHelper(heap.base, heap.cur_offset, alignment);
+    cur_heap_size = (uintptr_t)cur_chunk_base - (uintptr_t)heap.base;
 
-  // jemalloc 4.5.0 man: "If chunk is not NULL, the returned pointer must be
-  // chunk on success or NULL on error"
-  if (chunk && chunk != cur_chunk_base) {
-    pthread_mutex_unlock(&heap.alloc_lock);
-    return NULL;
-  }
+    // If there's not enough space on the heap for this allocation, return NULL
+    if (size > heap.size - cur_heap_size) {
+      pthread_mutex_unlock(&heap.alloc_lock);
+      return NULL;
+    }
 
-  cur_heap_size = (uintptr_t)cur_chunk_base - (uintptr_t)heap.base;
+    // Update the current pointer, now that we've past any early returns.
+    heap.cur_offset = cur_heap_size + size;
 
-  // If there's not enough space on the heap for this allocation, return NULL
-  if (size > heap.size - cur_heap_size) {
+    // now that cur_heap_offset is updated, we can unlock
     pthread_mutex_unlock(&heap.alloc_lock);
-    return NULL;
-  }
+  } else {
```

Comment:
How confident are you that we don't need to use the alloc_lock here? Is it safe for `chpl_comm_regMemAlloc()` and  `chpl_comm_regMemPostAlloc()` be called concurrently?

Reply:
Libhugetlbfs can back and/or replace malloc, so it should be thread-safe.  And the kernel code that handles hugepages (touch, etc.) certainly should be.  The comm layer would need to protect concurrent access to its data structures for managing registrations and so on, but the only implementation (comm=ugni, so far) does do so.  So yes, this should be safe.